### ISSUE TITLE
performance on large files

### ISSUE THREAD

**Issue by anson-vandoren** [srobinson2018/log-collection#4]
When requesting (for example) 1 million lines from a very large file, the service works, but takes quite a while to process the file. It's expected that it takes a while for this much, but it would be a better user experience if the results could be piped back to the requestor as they were being read in. This could also allow for lower memory usage, since the full 1 million lines do not need to be kept in memory before sending them.

**Comment by srobinson2018** [comment]
Absolutely! Streaming the data is definitely the more correct way to implement this type of application. Depending on the architecture of the platform, we could use websockets directly to a frontend app, a message bus (like Kafka), or even webhooks if we are integrating with an external consumer.

Given the (assumed) scope of this project, I am going to investigate a bit into using an http `keep-alive` header in order to handle multiple responses for a single http request. And then have the lines returned in batches.

**Comment by anson-vandoren** [comment]
@srobinson2018 I definitely agree that websockets and/or any sort of message bus is overkill here. I'm not as familiar with Go as some other languages, but I believe the http lib allows you to manually call a flush. In combination with a `Transfer-Encoding: chunked` header, this should be a method to push chunks of lines as you have them, instead of holding them all to the end.

**Comment by srobinson2018** [comment]
My recent changes added a streaming option that will stream the log results via a `100 - Continue`. I am not totally sure if this is the _best_ way to stream data via HTTP but it was the route I went with in my very quick search for the correct response code and headers.

This solution is a bit incomplete as I would ideally also have a buffered writer that would flush when full and/or on an interval (and then again at completion of processing the log file). What I implemented is writing each line as it is processed which slows it down pretty significantly for larger files.

Try it out with `writerType=stream` and let me know what you think!

Example:
```
curl "http://localhost:2022/logs?fileName=install.log&lineLimit=100000&grep=&format=&bufferSize=1&writerType=stream"
```
